sys/linux/fs_ioctl_btrfs.txt

```diff
@@ -1,7 +1,426 @@
-# Copyright 2018 syzkaller project authors. All rights reserved.
+# Copyright 2020 syzkaller project authors. All rights reserved.
 # Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
 
 include <uapi/linux/btrfs.h>
+include <uapi/linux/btrfs_tree.h>
+include <linux/refcount.h>
+include <linux/spinlock_types.h>
+include <linux/types.h>
+include <linux/blk_types.h>
+include <linux/wait.h>
+include <fs/btrfs/compression.h>
 
-# TODO: add the ioctls
-_ = BTRFS_IOC_SNAP_CREATE, BTRFS_IOC_DEFRAG, BTRFS_IOC_RESIZE, BTRFS_IOC_SCAN_DEV, BTRFS_IOC_TRANS_START, BTRFS_IOC_TRANS_END, BTRFS_IOC_SYNC, BTRFS_IOC_CLONE, BTRFS_IOC_ADD_DEV, BTRFS_IOC_RM_DEV, BTRFS_IOC_BALANCE, BTRFS_IOC_CLONE_RANGE, BTRFS_IOC_SUBVOL_CREATE, BTRFS_IOC_SNAP_DESTROY, BTRFS_IOC_FILE_EXTENT_SAME
+#BTRFS_IOC_SET_RECEIVED_SUBVOL_32 and BTRFS_IOC_SEND_32 are defined in <fs/btrfs/ioctl.c>, which I don't think should be included
+
+ioctl$BTRFS_IOC_SNAP_CREATE(fd fd, cmd const[BTRFS_IOC_SNAP_CREATE], arg ptr[in, btrfs_ioctl_vol_args])
+ioctl$BTRFS_IOC_SNAP_CREATE_V2(fd fd, cmd const[BTRFS_IOC_SNAP_CREATE_V2], arg ptr[in, btrfs_ioctl_vol_args_v2])
+ioctl$BTRFS_IOC_SUBVOL_CREATE(fd fd, cmd const[BTRFS_IOC_SUBVOL_CREATE], arg ptr[in, btrfs_ioctl_vol_args])
+ioctl$BTRFS_IOC_SUBVOL_CREATE_V2(fd fd, cmd const[BTRFS_IOC_SUBVOL_CREATE_V2], arg ptr[in, btrfs_ioctl_vol_args_v2])
+ioctl$BTRFS_IOC_SNAP_DESTROY(fd fd, cmd const[BTRFS_IOC_SNAP_DESTROY], arg ptr[in, btrfs_ioctl_vol_args])
+ioctl$BTRFS_IOC_SNAP_DESTROY_V2(fd fd, cmd const[BTRFS_IOC_SNAP_DESTROY_V2], arg ptr[in, btrfs_ioctl_vol_args_v2])
+ioctl$BTRFS_IOC_SUBVOL_GETFLAGS(fd fd, cmd const[BTRFS_IOC_SUBVOL_GETFLAGS], arg ptr[out, int64])
+ioctl$BTRFS_IOC_SUBVOL_SETFLAGS(fd fd, cmd const[BTRFS_IOC_SUBVOL_SETFLAGS], arg ptr[in, flags[btrfs_subvol_flags, int64]])
+ioctl$BTRFS_IOC_DEFAULT_SUBVOL(fd fd, cmd const[BTRFS_IOC_DEFAULT_SUBVOL], arg ptr[in, int64])
+ioctl$BTRFS_IOC_DEFRAG(fd fd, cmd const[BTRFS_IOC_DEFRAG], arg const[0])
+ioctl$BTRFS_IOC_DEFRAG_RANGE(fd fd, cmd const[BTRFS_IOC_DEFRAG_RANGE], arg ptr[in, btrfs_ioctl_defrag_range_args])
+ioctl$BTRFS_IOC_RESIZE(fd fd, cmd const[BTRFS_IOC_RESIZE], arg ptr[in, btrfs_ioctl_vol_args])
+ioctl$BTRFS_IOC_ADD_DEV(fd fd, cmd const[BTRFS_IOC_ADD_DEV], arg ptr[in, btrfs_ioctl_vol_args])
```

Comment:
Yes it definitely make sense to describe them differently. However I'm having trouble finding a way to describe them in syzkaller, as we need to generate strings of certain format (for example, in resize, we need [<devid>:][+/-]<size>[kKmMgGtTpPeE]|[<devid>:]max ). Any suggestion?

Reply:
If we need such a string, it can be represented as a struct along the lines of:
```
foo {
  # either ':' or nothing
  colon1 optional[const[':', int8]]
  # '+' or '-' or nothing
  plus_minus optional[flags[plus_minus, int8]]
  kKmM optional[flags[all_these_kKmM, int8]]
  max stringnoz["max"]
} [packed]

plus_minus = '+', '-'
```

Reply:
Thanks! This is extremely helpful. Just curious why max would be stringnoz instead of string?

Reply:
Unresolving, so we don't lose track of this improvement.